### PR TITLE
WIP: Add support for building on x86/64 FreeBSD host

### PR DESCRIPTION
**Pull request by rozhuk-im**
Allow to build OpenWRT on FreeBSD.
Tested on FreeBSD 12.1 amd64, with clang 10.
Image for D-Link DIR-510L build OK and work on h/w.

### CODE REVIEW

**Comment by diizzyy** [comment]
While I think it's a nice effort you're most likely going to run into countless of issues and I actually removed the FreeBSD stuff because it was broken. A lot of packages and whatnot makes a lot of assumptions that breaks on other platforms than Linux/GNU-land. You're much better off using bhyve and running Alpine/Debian or whatever you prefer and it'll cause you a lot less issues.

As for the patch itself,  no sign-off (hard requirement) and you'll most likely need to split this up into separate commits.

Personally I don't think you'll get this merged simply because it's not in any CI and there's little to no interest into adding one.

**Comment by rozhuk-im** [comment]
@diizzyy I can not agree with you.
I already have ubuntu in vbox to build owrt. This is terrible then I need to fix some thing in owrt:
host system: edit, git add, git commit --amend, git push -f
linux system: git reset, git pull and then build
This is too many things for one build. I want: edit, buld - without git and switch between terminals.

Hardcoding to allow OpenWRT be build only on few linux distros and only with gcc - bad practice.
I do not ask to support FreeBSD, I do not ask that all packages be build for FreeBSD - just let staff that really required to FreeBSD users to be work.

**Comment by ynezz** [comment]
> Personally I don't think you'll get this merged 

FYI few of us already tried to make it working, but as always got distracted by other stuff :) So I welcome this effort and would like to have FreeBSD supported as monocultures are bad.

> simply because it's not in any CI and there's little to no interest into adding one

You never know https://builds.sr.ht/~aparcar/job/235519

**Comment by rozhuk-im** [comment]
@ynezz Thanks, please wait a little, am working on build on clean system :)

**Comment by ynezz** [comment]
> please wait a little

No worries, there is some work ahead, for the start:

https://openwrt.org/submitting-patches#describe_your_changes
https://openwrt.org/submitting-patches#separate_your_changes

This huge commits needs to be split into smaller/separate units/commits with proper description, usually answer to "Why?" is this change needed (including the error output you're fixing or such) etc.

File: tools/cmake/patches/140-freebsd-build.fix.patch [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
++#include <sys/types.h>
+ 
+ #if defined(__GNUC__) && defined(__GNUC_MINOR__)
+ #   define TUKLIB_GNUC_REQ(major, minor) \
+@@ -126,7 +128,7 @@
+ 		(((uint16_t)(num) << 8) | ((uint16_t)(num) >> 8))
+ #endif
+ 
+-#ifndef bswap32
++#if !defined(bswap32) || defined(BSD)
```
Comment (sarog) [inline tools/cmake/patches/140-freebsd-build.fix.patch:17]: Consider replacing `defined(BSD)` with `defined(__FreeBSD__)`  instead.

File: tools/cmake/patches/140-freebsd-build.fix.patch [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ #	define bswap32(num) \
+ 		( (((uint32_t)(num) << 24)                       ) \
+ 		| (((uint32_t)(num) <<  8) & UINT32_C(0x00FF0000)) \
+@@ -134,7 +136,7 @@
+ 		| (((uint32_t)(num) >> 24)                       ) )
+ #endif
+ 
+-#ifndef bswap64
++#if !defined(bswap64) || defined(BSD)
```
Comment (sarog) [inline tools/cmake/patches/140-freebsd-build.fix.patch:26]: Consider replacing `defined(BSD)` with `defined(__FreeBSD__)`  instead.

**Review by sarog — COMMENTED** [review]
A few minor suggestions.

File: tools/cmake/patches/140-freebsd-build.fix.patch (hunk 2 above)
Comment (sarog) [reply tools/cmake/patches/140-freebsd-build.fix.patch:26]: Same suggestions apply for `scripts/mkhash.c`, however consider reviewing [this patch](https://bugs.openwrt.org/index.php?do=details&task_id=971&order=category&sort=desc) posted ~3 years ago.

**Comment by diizzyy** [comment]
@ynezz 
Who is going to maintain it (actually run FreeBSD)? There's going to quite a bit of patching required to get a lot of things working properly due to "Linux-ism" (see FreeBSDs ports tree) which is going to add more patches to the tree and that itself seems to be less than ideal. I doubt it's going to "worth the effort"/ be well maintained just by looking at history as it didn't even compile a basic image without jumping through hoops and on top of that packages will also fail due to the above. I doubt most maintainers are interested in support another OS they don't run.

File: tools/mtd-utils/include/linux/types.h [hunk 3]
```diff
@@ -10,7 +12,9 @@ typedef uint64_t __u64;
 typedef __u16 __le16;
 typedef __u32 __le32;
 typedef __u64 __le64;
+#ifndef BSD
```
Comment (sarog) [inline tools/mtd-utils/include/linux/types.h:15]: Replacing `BSD` with `__FreeBSD__` eliminates the need to include `sys/param.h` and `sys/types.h` above.

**Comment by rozhuk-im** [comment]
@diizzyy You miss one detail: FreeBSD ports tree have patches:
1. That is not yet upstreamed
2. That required to build and run on FreeBSD.
There it no need in "run" patches from FreeBSD.

I do not think that packages maintainers will get even minimal count of patches.
To build all that I use on FreeBSD only one line in one package must be changed: https://github.com/openwrt/packages/pull/12516

Python installed as dep for other ports.

**Comment by diizzyy** [comment]
There's a lot more due to host-builds and when all packages (not just a handful) in main repo builds it might be worth considering including it although I can only imagine how much breakage you'd see in the package repo.

**Comment by adschm** [comment]
This has formal issues: commit title needs prefix (and is probably too generic), commit message is missing.

Since nobody has really looked at it for almost one year, I wonder whether we should keep it open at all at this point?

**Comment by rozhuk-im** [comment]
This PR updates time to time, I do builds for myself.

**Comment by adschm** [comment]
Okay, so this not really meant to be included anymore anyway.

**Pull request closed without merging**